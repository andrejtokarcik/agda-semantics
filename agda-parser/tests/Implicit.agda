
-- This module introduces implicit arguments.

module Implicit where

-- In Agda you can omit things that the type checker can figure out for itself.
-- This is a crucial feature in a monomorphic language, since you would
-- otherwise be overwhelmed by type arguments.

-- Let's revisit the identity function from 'Introduction.Basics'.

id' : (A : Set) -> A -> A
id' A x = x

-- Since Agda is monomorphic we have to take the type A as an argument. So when
-- using the identity function we have to provide the type explicitly.

data Nat : Set where
  zero : Nat
  suc  : Nat -> Nat

one : Nat
one = id' Nat (suc zero)

-- Always having to provide the type argument to the identity function would of
-- course be very tedious, and seemingly unnecessary. We would expect the type
-- checker to be able to figure out what it should be by looking at the second
-- argument. And indeed the type checker can do this.

-- One way of indicating that the type checker will have figure something out
-- for itself is to write an underscore (_) instead of the term. So when
-- applying the identity function we can write

two : Nat
two = id' _ (suc one)

-- Now the type checker will try to figure out what the _ should be, and in
-- this case it will have no problems doing so. If it should fail to infer the
-- value of an _ it will issue an error message.

-- In the case of the identity function we expect the type argument to always
-- be inferrable so it would be nice if we could avoid having to write the _.
-- This can be achieved by declaring the first argument as an implicit
-- argument.

id : {A : Set} -> A -> A        -- implicit arguments are enclosed in curly braces
id x = x  -- now we don't have to mention A in the left-hand side

three : Nat
three = id (suc two)

-- If, for some reason, an implicit argument cannot be inferred it can be given
-- explicitly by enclosing it in curly braces :

four : Nat
four = id {Nat} (suc three)

-- To summarise we give a bunch of possible variants of the identity function
-- and its use.

-- Various definitions of the identity function. Definitions 0 through 3 are
-- all equivalent, as are definitions 4 through 6.

id0 : (A : Set) -> A -> A
id0 A x = x

id1 : (A : Set) -> A -> A
id1 _ x = x     -- in left-hand sides _ means "don't care"

id2 : (A : Set) -> A -> A
id2 = \A x -> x

id3 = \(A : Set)(x : A) -> x  -- the type signature can be omitted for definitions
                          -- of the form x = e if the type of e can be
                          -- inferred.

id4 : {A : Set} -> A -> A
id4 x = x

id5 : {A : Set} -> A -> A
id5 {A} x = x

id6 : {A : Set} -> A -> A
id6 = \x -> x

id7 = \{A : Set}(x : A) -> x

id8 : {A : Set} -> A -> A
id8 = \{A} x -> x        -- this doesn't work since the type checker assumes
                            -- that the implicit A has been has been omitted in
                            -- the left-hand side (as in id6).

-- Various uses of the identity function.
zero0 = id0 Nat zero
zero1 = id0 _   zero  -- in right-hand sides _ means "go figure"

zero2 = id4       zero
zero3 = id4 {Nat} zero
zero4 = id4 {_}   zero  -- This is equivalent to zero2, but it can be useful if
                        -- a function has two implicit arguments and you need
                        -- to provide the second one (when you provide an
                        -- implicit argument explicitly it is assumed to be the
                        -- left-most one).

-- In this module we have looked at implicit arguments as a substitute for
-- polymorphism. The implicit argument mechanism is more general than that and
-- not limited to inferring the values of type arguments. For more information
-- on implicit arguments see, for instance
--  'Introduction.Data.ByRecursion'

