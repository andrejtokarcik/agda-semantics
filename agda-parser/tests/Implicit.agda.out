<T>
    <k>
        .K
    </k>
    <mgu>
        subst(.KList)
    </mgu>
    <tenvRestore>
        .List
    </tenvRestore>
    <metadata>
        <range>
            .K
        </range>
    </metadata>
    <con>
        Data ( "Nat" ) |-> l ( ListItem(Con ( "zero" )) ListItem(Con ( "suc" )
          ) )
    </con>
    <funtenvs>
        Fun ( "zero4" ) |-> m ( .Map )
    </funtenvs>
    <metavars>
        .Map
    </metavars>
    <fenv>
        Fun ( "four" ) |-> ((Fun ( "id" ) { Data ( "Nat" ) }) (Con ( "suc" ) 
          Fun ( "three" )))
        Fun ( "id" ) |-> FunClauses ( ListItem(((#symFunClauses(40) { #symExpr
          (39) }) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id'" ) |-> FunClauses ( ListItem(((#symFunClauses(11) #symExpr(
          10)) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id0" ) |-> FunClauses ( ListItem(((#symFunClauses(68) #symExpr(
          67)) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id1" ) |-> FunClauses ( ListItem(((#symFunClauses(78) #symExpr(
          77)) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id2" ) |-> \ ( #symExpr(101) : #symExpr(79) ) -> \ ( Var ( "x" 
          ) : #symExpr(82) ) -> Var ( "x" )
        Fun ( "id3" ) |-> \ ( #symExpr(101) : Set ( 0 ) ) -> \ ( Var ( "x" ) :
           #symExpr(101) ) -> Var ( "x" )
        Fun ( "id4" ) |-> FunClauses ( ListItem(((#symFunClauses(96) { 
          #symExpr(95) }) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id5" ) |-> FunClauses ( ListItem(((#symFunClauses(102) { 
          #symExpr(101) }) Var ( "x" )) = Var ( "x" )) )
        Fun ( "id6" ) |-> FunClauses ( ListItem((#symFunClauses(108) { 
          #symExpr(106) }) = \ ( Var ( "x" ) : #symExpr(103) ) -> Var ( "x" ))
           )
        Fun ( "id7" ) |-> (\ { Var ( "A" ) : Set ( 0 ) } -> \ ( Var ( "x" ) : 
          Var ( "A" ) ) -> Var ( "x" ) { #symExpr(112) })
        Fun ( "id8" ) |-> FunClauses ( ListItem((#symFunClauses(123) { 
          #symVariable(121) }) = (\ { Var ( "A" ) : Set ( 0 ) } -> \ ( Var ( 
          "x" ) : #symExpr(118) ) -> Var ( "x" ) { #symVariable(121) })) )
        Fun ( "one" ) |-> ((Fun ( "id'" ) Data ( "Nat" )) (Con ( "suc" ) Con (
           "zero" )))
        Fun ( "three" ) |-> ((Fun ( "id" ) { Data ( "Nat" ) }) (Con ( "suc" ) 
          Fun ( "two" )))
        Fun ( "two" ) |-> ((Fun ( "id'" ) Data ( "Nat" )) (Con ( "suc" ) Fun (
           "one" )))
        Fun ( "zero0" ) |-> ((Fun ( "id0" ) Data ( "Nat" )) Con ( "zero" ))
        Fun ( "zero1" ) |-> ((Fun ( "id0" ) Data ( "Nat" )) Con ( "zero" ))
        Fun ( "zero2" ) |-> ((Fun ( "id4" ) { Data ( "Nat" ) }) Con ( "zero" )
          )
        Fun ( "zero3" ) |-> ((Fun ( "id4" ) { Data ( "Nat" ) }) Con ( "zero" )
          )
        Fun ( "zero4" ) |-> ((Fun ( "id4" ) { Data ( "Nat" ) }) Con ( "zero" )
          )
    </fenv>
    <tenv>
        .Map
    </tenv>
    <sig>
        Fun ( "four" ) |-> Data ( "Nat" )
        Fun ( "id" ) |-> { #symExpr(67) : Set ( 0 ) } -> ( #symVariable(56) : 
          #symExpr(67) ) -> #symExpr(67)
        Fun ( "id'" ) |-> ( #symExpr(10) : Set ( 0 ) ) -> ( #symVariable(30) :
           #symExpr(10) ) -> #symExpr(10)
        Fun ( "id0" ) |-> ( #symExpr(67) : Set ( 0 ) ) -> ( #symVariable(142) 
          : #symExpr(67) ) -> #symExpr(67)
        Fun ( "id1" ) |-> ( #symExpr(82) : Set ( 0 ) ) -> ( #symVariable(75) :
           #symExpr(82) ) -> #symExpr(82)
        Fun ( "id2" ) |-> ( #symExpr(82) : Set ( 0 ) ) -> ( Var ( "x" ) : 
          #symExpr(82) ) -> #symExpr(82)
        Fun ( "id3" ) |-> ( #symExpr(101) : Set ( 0 ) ) -> ( Var ( "x" ) : 
          #symExpr(101) ) -> #symExpr(101)
        Fun ( "id4" ) |-> { #symExpr(101) : Set ( 0 ) } -> ( #symVariable(167)
           : #symExpr(101) ) -> #symExpr(101)
        Fun ( "id5" ) |-> { #symExpr(101) : Set ( 0 ) } -> ( #symVariable(99) 
          : #symExpr(101) ) -> #symExpr(101)
        Fun ( "id6" ) |-> { Var ( "A" ) : Set ( 0 ) } -> ( Var ( "x" ) : Var (
           "A" ) ) -> Var ( "A" )
        Fun ( "id7" ) |-> ( Var ( "x" ) : #symExpr(112) ) -> #symExpr(112)
        Fun ( "id8" ) |-> { Var ( "A" ) : Set ( 0 ) } -> ( Var ( "x" ) : Var (
           "A" ) ) -> Var ( "A" )
        Fun ( "one" ) |-> Data ( "Nat" )
        Fun ( "three" ) |-> Data ( "Nat" )
        Fun ( "two" ) |-> Data ( "Nat" )
        Fun ( "zero0" ) |-> Data ( "Nat" )
        Fun ( "zero1" ) |-> Data ( "Nat" )
        Fun ( "zero2" ) |-> Data ( "Nat" )
        Fun ( "zero3" ) |-> Data ( "Nat" )
        Fun ( "zero4" ) |-> Data ( "Nat" )
        Data ( "Nat" ) |-> Set ( 0 )
        Con ( "suc" ) |-> ( #symVariable(53) : Data ( "Nat" ) ) -> Data ( 
          "Nat" )
        Con ( "zero" ) |-> Data ( "Nat" )
    </sig>
    <ncat>
        Name ( "Nat" ) |-> Data ( "Nat" )
        Name ( "four" ) |-> Fun ( "four" )
        Name ( "id" ) |-> Fun ( "id" )
        Name ( "id'" ) |-> Fun ( "id'" )
        Name ( "id0" ) |-> Fun ( "id0" )
        Name ( "id1" ) |-> Fun ( "id1" )
        Name ( "id2" ) |-> Fun ( "id2" )
        Name ( "id3" ) |-> Fun ( "id3" )
        Name ( "id4" ) |-> Fun ( "id4" )
        Name ( "id5" ) |-> Fun ( "id5" )
        Name ( "id6" ) |-> Fun ( "id6" )
        Name ( "id7" ) |-> Fun ( "id7" )
        Name ( "id8" ) |-> Fun ( "id8" )
        Name ( "one" ) |-> Fun ( "one" )
        Name ( "suc" ) |-> Con ( "suc" )
        Name ( "three" ) |-> Fun ( "three" )
        Name ( "two" ) |-> Fun ( "two" )
        Name ( "zero" ) |-> Con ( "zero" )
        Name ( "zero0" ) |-> Fun ( "zero0" )
        Name ( "zero1" ) |-> Fun ( "zero1" )
        Name ( "zero2" ) |-> Fun ( "zero2" )
        Name ( "zero3" ) |-> Fun ( "zero3" )
        Name ( "zero4" ) |-> Fun ( "zero4" )
    </ncat>
</T>
